GVFS/GVFS.Virtualization/FileSystemCallbacks.cs

```diff
@@ -296,6 +296,48 @@ public EventMetadata GetAndResetHeartBeatMetadata(out bool logToFile)
             return metadata;
         }
 
+        public bool DehydrateFolder(string relativePath)
+        {
+            List<IPlaceholderData> removedPlaceholders;
+            List<string> removedModifiedPaths;
+            bool successful = false;
+
+            try
+            {
+                removedPlaceholders = this.placeholderDatabase.RemoveStartingWith(relativePath);
+                removedModifiedPaths = this.modifiedPaths.RemoveAllEntriesForFolder(relativePath);
+                FileSystemResult result = this.fileSystemVirtualizer.DehydrateFolder(relativePath);
+                successful = result.Result == FSResult.Ok;
+            }
+            catch (Exception ex)
+            {
+                EventMetadata metadata = this.CreateEventMetadata(relativePath, ex);
+                this.context.Tracer.RelatedError(metadata, $"{nameof(this.DehydrateFolder)} threw and exception");
+                return false;
+            }
+
+            if (!successful)
+            {
+                if (removedPlaceholders != null)
+                {
+                    foreach (IPlaceholderData data in removedPlaceholders)
+                    {
+                        this.placeholderDatabase.AddPlaceholderData(data);
+                    }
+                }
+
+                if (removedModifiedPaths != null)
+                {
+                    foreach (string modifiedPath in removedModifiedPaths)
+                    {
+                        this.modifiedPaths.TryAdd(modifiedPath, isFolder: modifiedPath.EndsWith(GVFSConstants.GitPathSeparatorString), isRetryable: out bool isRetryable);

```

Comment:
Do we need to check the return value of `TryAdd`?  Should it always succeed?

Reply:
I'm not sure there is anything we can do if it is false.  It was already logged in the `TryAdd`.  We could retry a few times when `isRetryable` is true.  I guess the other option is to add it to the background queue and let it handle adding?  That seems a little hacky.  Let me know your thoughts.